File: js/2239-스도쿠.js
const readline = require('readline');

(async () => {
  let rl = readline.createInterface({ input: process.stdin });

  let count = 0;

  const sdocu = [];
  for await (const line of rl) {
    sdocu.push(line.split('').map(Number));
    count++;

    if (count === 9) {
      rl.close();
    }
  }

  const check = (x, y, num) => {
    if (sdocu[x].includes(num)) {
      return false;
    }
    for (let i = 0; i < 9; i++) {
      if (sdocu[i][y] === num) {
        return false;
      }
    }
    for (let i = x * 3; i < (x + 1) * 3 - 1; i++) {
      for (let j = y * 3; j < (y + 1) * 3 - 1; j++) {
        if (sdocu[i][j] === num) {
          return false;
        }
      }
    }
    return true;
  };

  const Back = (index) => {
    // 탈출 조건
    if (index === arr.length - 1) {
      sdocu.forEach((list) => console.log(list.join('')));
      return;
    }

    for (let i = 1; i < 3; i++) {
      const [x, y] = arr[index];
      console.log(x, y, i);
      console.log(Math.floor(x / 3), Math.floor(y / 3), i);
      if (check(Math.floor(x / 3), Math.floor(y / 3), i)) {
        // console.log(x, y, i);
        sdocu[x][y] = i;
        Back(index + 1);
        sdocu[x][y] = 0;
      }
    }
  };

  // 스도쿠 숫자 넣어야하는 좌표 다 찾기
  const arr = [];
  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      if (sdocu[i][j] === 0) {
        arr.push([i, j]);
      }
    }
  }

  console.log(arr);
  Back(0);
  process.exit();
})();

File: 2239-스도쿠.py
import sys 
input = sys.stdin.readline

sdocu = [list(map(int, input().rstrip())) for _ in range(9)]

def kan(x,y,n):
    a, b = x // 3, y // 3

    for i in range(3*a, 3*a+3):
        for j in range(3*b, 3*b+3):
            if sdocu[i][j] == n:
                return False
    return True

def sero(y,n):
    for i in range(9):
        if sdocu[i][y] == n:
            return False
    return True

def solveSdocu(index):
    # 반환 조건
    if index == len(arr): 
        for i in range(9):
            print(''.join(map(str,sdocu[i])))
        exit()
    
    # 값 넣을 좌표
    x, y = arr[index][0], arr[index][1]

    # sdocu 해결
    for i in range(1,10):
        # 해당 숫자가 들어갈 수 있는지 체크
        if i not in sdocu[x] and kan(x,y,i) and sero(y,i):
            sdocu[x][y] = i
            solveSdocu(index+1)
            sdocu[x][y] = 0

arr = []
for i in range(9):
    for j in range(9):
        if sdocu[i][j] == 0:
            arr.append([i,j])

solveSdocu(0)